x-pack/plugins/siem/public/hooks/use_add_to_timeline.tsx

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import d3 from 'd3';
+import { useCallback } from 'react';
+import { DraggableId, FluidDragActions, Position, SensorAPI } from 'react-beautiful-dnd';
+
+import { IS_DRAGGING_CLASS_NAME } from '../components/drag_and_drop/helpers';
+import { HIGHLIGHTED_DROP_TARGET_CLASS_NAME } from '../components/timeline/data_providers/empty';
+import { EMPTY_PROVIDERS_GROUP_CLASS_NAME } from '../components/timeline/data_providers/providers';
+
+let _sensorApiSingleton: SensorAPI;
+
+/**
+ * This hook is passed (in an array) to the `sensors` prop of the
+ * `react-beautiful-dnd` `DragDropContext` component. Example:
+ *
+ * ```
+       <DragDropContext onDragEnd={onDragEnd} sensors={[useAddToTimelineSensor]}>
+        {children}
+      </DragDropContext>*
+ * ```
+ *
+ * As a side effect of registering this hook with the `DragDropContext`,
+ * the `SensorAPI` singleton is initialized. This singleton is used
+ * by the `useAddToTimeline` hook.
+ */
+export const useAddToTimelineSensor = (api: SensorAPI) => {
+  _sensorApiSingleton = api;
+};
+
+/**
+ * Returns the position of the specified element
+ */
+const getPosition = (element: Element): Position => {
+  const rect = element.getBoundingClientRect();
+
+  return { x: rect.left, y: rect.top };
+};
+
+/**
+ * Returns the position of one of the following timeline drop targets
+ * (in the following order of preference):
+ * 1) The "Drop anything highlighted..." drop target
+ * 2) The persistent "empty" data provider group drop target
+ * 3) `null`, because none of the above targets exist (an error state)
+ */
+export const getDropTargetCoordinate = (): Position | null => {
+  // The placeholder in the "Drop anything highlighted here to build an OR query":
+  const highlighted = document.querySelector(`.${HIGHLIGHTED_DROP_TARGET_CLASS_NAME}`);
+
+  if (highlighted != null) {
+    return getPosition(highlighted);
+  }
+
+  // If at least one provider has been added to the timeline, the "Drop anything
+  // highlighted..." drop target won't be visible, so we need to drop into the
+  // empty group instead:
+  const emptyGroup = document.querySelector(`.${EMPTY_PROVIDERS_GROUP_CLASS_NAME}`);
+
+  if (emptyGroup != null) {
+    return getPosition(emptyGroup);
+  }
+
+  return null;
+};
+
+/**
+ * Returns the coordinates of the specified draggable
+ */
+export const getDraggableCoordinate = (draggableId: DraggableId): Position | null => {
+  // The placeholder in the "Drop anything highlighted here to build an OR query":
+  const draggable = document.querySelector(`[data-rbd-draggable-id="${draggableId}"]`);
```

Comment:
Nothing to do here but I was wondering when we refactor the timeline if we can try to remember this querySelector so we tried to replace them by`ref` from react. it might be worth to try that.